libc/vga/rlinit-vesa.S

```diff
@@ -0,0 +1,611 @@
+│ This is free and unencumbered software released into the public domain.      │
+│                                                                              │
+│ Anyone is free to copy, modify, publish, use, compile, sell, or              │
+│ distribute this software, either in source code form or as a compiled        │
+│ binary, for any purpose, commercial or non-commercial, and by any            │
+│ means.                                                                       │
+│                                                                              │
+│ In jurisdictions that recognize copyright laws, the author or authors        │
+│ of this software dedicate any and all copyright interest in the              │
+│ software to the public domain. We make this dedication for the benefit       │
+│ of the public at large and to the detriment of our heirs and                 │
+│ successors. We intend this dedication to be an overt act of                  │
+│ relinquishment in perpetuity of all present and future rights to this        │
+│ software under copyright law.                                                │
+│                                                                              │
+│ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,              │
+│ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF           │
+│ MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.       │
+│ IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR            │
+│ OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,        │
+│ ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR        │
+│ OTHER DEALINGS IN THE SOFTWARE.                                              │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "ape/relocations.h"
+#include "libc/macros.internal.h"
+#include "libc/runtime/mman.internal.h"
+#include "libc/vga/vga.internal.h"
+
+#define VGA_PREFER_GRAPH_HEIGHT \
+	(VGA_PREFER_TTY_HEIGHT * VGA_ASSUME_CHAR_HEIGHT_PX)
+#define VGA_PREFER_GRAPH_WIDTH  \
+	(VGA_PREFER_TTY_WIDTH * VGA_ASSUME_CHAR_WIDTH_PX)
+#define MAX_VESA_MODES_TO_TRACK 64
+
+//	Mode information data structure, used internally.
+	.set	"mi::type",0
+	.set	"mi::bpp",1
+	.set	"mi::width",2
+	.set	"mi::height",4
+	.set	"mi::mode_num",6
+	.set	"mi::stride",8
+	.set	"mi::fb",10
+	.set	"mi::sizeof",14
+
+//	Routine to activate additional VESA functionality if the user holds
+//	down a magic key, when booting from bare metal.  Currently this just
+//	dumps a list of all usable VESA VBE video modes on the console.
+//
+//	TODO: allow user to select a video mode and switch to it
+//	TODO: implement character drawing in graphics modes
+//
+//	@return	CF = 0 if we decided to set a new video mode, CF = 1 otherwise
+
+	.real
+	.code16
+_rlinit_vesa:
+	push	%es
+	pushw	$0
+	testb	$0b00000011,0x0417	# read keyboard shift state (as
```

Comment:
I've re-enabled the Wiki feature on this repository. Knowledge like a magic shift key for VGA config would be a good thing to have documented there! Put anything on the wiki you want!